### DISCUSSION TITLE
Need help as our application is still using the /v2 endpoint to monitor etcd. But /v2 - API is deprecated now.

### DISCUSSION THREAD

**Discussion by bipinchandra1** [etcd-io/etcd#16076] [Q&A] [answered]
We monitor following stats from these two endpoints - 

**/v2/stats/self**
- requests_received
- bytes_per_sec_received
- requests_per_sec_received
- requests_sent
- bytes_per_sec_sent
- requests_per_sec_sent

**/v2/stats/store**
- compare_and_swap_fail
- compare_and_swap_success
- compare_and_delete_fail
- compare_and_delete_success
- create_fail
- create_success
- delete_fail
- delete_success
- expire_count
- gets_fail
- gets_success
- sets_fail
- sets_success
- update_fail
- update_success
- watchers

We need help as how to migrate these stats to /v3 endpoint.

**Comment by jmhbnz** [comment]
Hey @bipinchandra1 - Thanks for your question, please refer to etcd v3 monitoring documentation in the first instance: 
 - https://etcd.io/docs/v3.5/op-guide/monitoring
 - https://etcd.io/docs/v3.5/metrics

**Reply by bipinchandra1** [reply]
Thanks @jmhbnz .
We are able to get all the stats from /metrics endpoint. It works same as v2 and v3.
But we could not get the corresponding stats related to /v2/stats/self and /v2/stats/store endpoints.
I have provided the complete list above which is possible in v2 but not in v3.
Need help.

**Reply by jmhbnz** [reply]
As linked in the metrics documentation for v3 there is a full list of metrics available: https://etcd.io/docs/v3.5/metrics/etcd-metrics-latest.txt 

At first glance I can see many of the things in your list for example:
- `watchers` --> `etcd_debugging_store_watchers`
- `gets_success` --> `etcd_debugging_store_reads_total`

Unless I am misunderstanding something about etcdv3 you will need to update your monitoring to use the new metrics available, following the operations guide documentation I linked above.

**Reply by bipinchandra1** [reply]
Thank you @jmhbnz .
I see only a few of these metrics are getting mapped. 
If I understood it correctly there would not be a direct mapping but we will have to re-implement the monitoring of etcd based on the new set of metrics available in the list you have shared.
Please confirm.

**Reply by bipinchandra1** [reply]
Following are the best I could do to map these. Many metrics are not available. Any idea how can we get the **not available** ones.  



/v2/stats/self | Possible mapping - v3 /metrics endpoints (wrt grpc)
-- | --
requests_received (self_recv_appendreq_cnt) | grpc_server_msg_received_total
bytes_per_sec_received (self_recv_bandwidth_rate) | etcd_network_client_grpc_received_bytes_total
requests_per_sec_received (self_recv_pkg_rate) | **Not available**
requests_sent (self_appendreq_cnt) | grpc_server_msg_sent_total
bytes_per_sec_sent (sendbandwidthrate) | etcd_network_client_grpc_sent_bytes_total
requests_per_sec_sent (self_send_pkg_rate) | **Not available**
 
    

/v2/stats/store |  Possible mapping - v3 /metrics endpoints
-- | --
compare_and_swap_fail | **Not available**
compare_and_swap_success | etcd_debugging_store_writes_total. [action = compareAndSwap]
compare_and_delete_fail | **Not available**
compare_and_delete_success | etcd_debugging_store_writes_total. [action = compareAndDelete]
create_fail | **Not available**
create_success | etcd_debugging_store_writes_total. [action = create]
delete_fail | **Not available**
delete_success | etcd_debugging_store_writes_total. [action = delete]
expire_count | etcd_debugging_store_expires_total
gets_fail | **Not available**
gets_success | etcd_debugging_store_reads_total. [action = get]
sets_fail | **Not available**
sets_success | etcd_debugging_store_writes_total [action = success]
update_fail | **Not available**
update_success | etcd_debugging_store_writes_total. [action = update]
watchers | etcd_debugging_store_watchers

Any idea how can we get the **not available** ones.

**Reply by jmhbnz** [reply, marked as answer]
Hey @bipinchandra1 - Rather than trying to use metrics because they existed in v2 my suggestion would be to step back and review what it is you are trying to understand or prove through monitoring etcd metrics.

There are fundamental differences between v2 and v3 that mean some of the old approaches are no longer relevant, refer to https://etcd.io/docs/v3.5/learning/api_guarantees to refresh on the kv api guarantees that etcd v3 provides.

Please follow the documented approaches for monitoring etcd v3 and if you find that the monitoring is insufficient to prove a specific type of activity or behavior of etcd then you're welcome at that point to open a feature request explaining in detail what you think is missing and why it is important.

**Reply by bipinchandra1** [reply]
Thank you @jmhbnz . Really helpful.
